sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -505,6 +505,9 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
         } else {
           e.copy(branches = branches.take(i).map(branch => (branch._1, elseValue)))
         }
+
+      case CaseWhen(Seq((cond, trueValue)), elseValue) =>
```

Comment:
Is it better we limit to `BooleanType` case? I.e.,

```scala
case cw @ CaseWhen(Seq((cond, trueValue)), elseValue) if cw.dataType == BooleanType =>
```

The reason is because mostly the further optimization comes from #29567, and it is for boolean type case only.

Or just rewrite it similarly like #29567?

```scala
case CaseWhen(Seq((cond, l @ Literal(null, _))), FalseLiteral) if !cond.nullable => ...
case CaseWhen(Seq((cond, l @ Literal(null, _))), TrueLiteral) if !cond.nullable => ...
```